modules/cluster-logging-collector-log-forwarding-about.adoc

```diff
@@ -142,3 +142,13 @@ spec:
 ** The `outputRefs` is the name of the output to use.
 ** Optional: A label to add to the logs.
 
+[discrete]
+[id="cluster-logging-external-fluentd"]
+== Fluentd log handling when the external log aggregator is unavailable
+
+If your external logging aggregator becomes unavailable and cannot receive logs, Fluentd continues to collect logs and stores them in a buffer. When the log aggregator becomes available, log forwarding resumes, including the buffered logs. The buffer can grow to 10% of the size of the disk in the server where the Fluentd node was scheduled. If the buffer fills completely, Fluentd stops collecting logs. {product-title} rotates the logs and deletes them. You cannot adjust the buffer size or add a persistent volume claim (PVC) to the Fluentd daemon set or pods.
+
+Each log forwarding output has two buffers: a main buffer where chunks are created and a retry buffer that holds any chunks that failed to be sent. If you have multiple log forwarding outputs, the maximum disk space for the buffers is divided among the buffers.
+
+For example, if a node is 120G, Fluentd can use 12G for buffers, which is 10%. If your cluster has four defined outputs, there are a total of eight buffers, with each buffer allocated 1.5G, which is the 12G allocated for the buffers divided by the eight buffers.
```

Comment:
I don't want to give the impression that this is a guaranteed behaviour for future versions. Buffer sizes, number of buffers etc. are internal details that may change without affecting the intended behaviour which is:

* On disconnect we will attempt to reconnect automatically.
* Logs are stored while disconnected and will be sent on reconnect.
* Long disconnects may result in loss of some logs.

It might be useful to have the current details somewhere but out of the main text so it's clear they are details that can change, not part of the guaranteed behavior.

For example I believe the large fluentd buffers are a mistake - the logs are already stored on disk, there's no value to maintaining large buffers in fluentd. But the observable behaviour is the same regardless of whether logs are stored in log files or in buffers during a disconnect.

Reply:
@sichvoge Do you have an opinion on what information we should present here?

Reply:
@mburke5678 What was the intent do put those information in that detail in the first place?

Reply:
@sichvoge I guess I wanted to give people some sense about how much space is available for buffering the logs and, thus, some sense about long they have to fix the problem with the aggregator before logs are lost.

Reply:
Yeah. If we are not confident on that data yet - we should not talk about it. 

@alanconway Why is it not configurable? I can see some customers that have different requirements on how long data is kept before loosing them.

Reply:
> Yeah. If we are not confident on that data yet - we should not talk about it.
> 
> @alanconway Why is it not configurable? I can see some customers that have different requirements on how long data is kept before loosing them.

Long answer, not relevant to docs!

Log data is *stored* in two places: /var/log on the node, and elastic (or whatever). Logs are available for as long as they exist in one of those two places.

Buffers in the forwarder are *not* for storing logs, they are for smoothing out temporary shortfall between data-in/data-out rates. Sizing those buffers is about variation in relative scrape/send rates, it has no bearing on user retention goals *or log loss*.

If logRate > forwardRate then we will lose, on average, (logRate - forwardRate)*t over time t, regardless of buffer sizes. That's a mathematical certainty. However big buffers vs. small buffers will change the *pattern* of log loss.

Our current huge buffers mean that we accumulate a huge backlog of old logs before we start to lose logs. That doesn't mean we lose less logs - it means we lose them in larger batches while we are chewing thru the backlog, and we deliver the old backed-up logs with high latency - they've been sitting in the buffer for ages. So we get slow delivery with occasional big gaps.

With short buffers, when we lose a chunk of logs we almost immediately move on to the next available chunk. We lose the same amount of logs in the long run, but we lose smaller chunks more often.  The logs we don't lose are delivered with low latency - they haven't been stuck in our buffers. So the user gets faster, more up-to-date logs with more frequent but smaller gaps.

I can't think of any use case where you would not prefer the latter pattern.

Reply:
@sichvoge @alanconway I removed those two paragraphs and the reference to the buffer filling at 10% of disk size:

_If your external logging aggregator becomes unavailable and cannot receive logs, Fluentd continues to collect logs and stores them in a buffer. When the log aggregator becomes available, log forwarding resumes, including the buffered logs. If the buffer fills completely, Fluentd stops collecting logs. {product-title} rotates the logs and deletes them. You cannot adjust the buffer size or add a persistent volume claim (PVC) to the Fluentd daemon set or pods._